Enable filter optimisation

Firstyear opened the pull request:
Supercedes https://github.com/389ds/389-ds-base/pull/5168

mreynolds389 commented on ldap/servers/slapd/back-ldbm/ldbm_search.c:
```diff
@@ -999,13 +1005,33 @@ build_candidate_list(Slapi_PBlock *pb, backend *be, struct backentry *e, const c
-        *candidates = subtree_candidates(pb, be, base, e, filter, managedsait,
-                                         lookup_returned_allidsp, &err);
+        /* Dup before optimise to prevent issues with aci and filter as intended vs optimised. */
+        filter_copy = slapi_filter_dup(filter);
```
This filter duping & freeing for every search concerns me from a performance & memory frag perspective.  I wonder if create_*_filter() functions could be modified to only dup the filter if the filter is actually modified.  Return NULL if not modified, and use the original filter, otherwise use the copy.  That would at least reduce this overhead for cases where the optimization is not needed.

Firstyear replied:
But there aren't many cases where this would happen. Create onelevel always has to add the parentid element. Create subtree always has to add the objectClass referral (unless we get the managedsait control, which practically no one sends). There additions were what caused the issue before. 

About the only way to potential min-max this to avoid alloc/dealloc would be to pull out the optimised inner, but then we loose the ability to optimise around parentid= and oc=referral. 

Even with the suggestion to have a filter as intended vs as optimised, we have to do an extra dup/free. So that's unavoidable unless we do some weird tracking of the inner filter that we optimise, and then the wrapper that we don't, and we free them separately or something. Which risks more memory corruption and weirdness. Generally this all gets very messy ....

mreynolds389 replied:
Sure, makes sense.  I was just hoping it could be improved, but doesn't look that way.  Ok, thanks for listening :-)  This fix is going through the entire IPA test suite at the moment.  And hopefully this week I can do some testing numbers.  

I know @tbordaz had some concerns that this fix will impact a feature where we skip acl evaluation of filter components.  He thought this might actually break that - which in turn would hurt performance (or worse?).  I asked him to go into detail and comment on this PR.  He hasn't had a chance to do so yet, but I'd like his concerns looked into (once they are accurately expressed).

In the meantime I hope to run some performance tests.  What filters do you suggest I test that will best showcase this feature's improvements?

Firstyear replied:
Wellllll after a walk I think there MIGHT be a way to do what you want ... so let me think a bit and come up with a new patch on top today.

The classic example is "(&(objectClass=person)(uid=foo))". But generally most of the filters that SSSD emits would benefit too, I remember one of the filters was pathologically bad, but I don't recall what it was now.


Since this is only a first pass, it'll only improve things like the above, but later we can make more interesting optimisations.

tbordaz commented on ldap/servers/slapd/back-ldbm/filterindex.c:
```diff
@@ -899,11 +903,12 @@ list_candidates(
              * If we encounter a zero length idl, we bail now because this can never
              * result in a meaningful result besides zero.
              */
+            sr->sr_flags |= SR_FLAG_MUST_APPLY_FILTER_TEST;
```
Not sure to understand, in such case we have no candidate. So why do we need to enforce the filter ?

Firstyear replied:
Because in the future we may add other optimisations to the OR case, that would still require filter eval when there are partial sets. Also remember, that provided the candidate set is less than test threshold, we early return and thus MUST do a filter test.

Firstyear replied:
Which by the way is the current behaviour, it's just that now we flag this in the filterindex.c code instead, because before it was flagged by lookup_returned_allidsp in onelevel_candidates().

tbordaz commented on ldap/servers/slapd/back-ldbm/ldbm_search.c:
```diff
@@ -1695,31 +1694,41 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
                                   "Bypassing filter test\n");
                     if (ACL_CHECK_FLAG) {
-                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
+                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
```
I am missing something, filter_intent is the original and optimized filter but it does not contain (objectclass=referral). If the candidate is a referral, this filter will not match and the referral will be skipped. ?

Firstyear replied:
we skip checking it against *aci* specifically. because the user may not have the aci to read referals. The same way the user may not have the aci to see parentid. But we still need to enforce this condition in execution ....

tbordaz commented on ldap/servers/slapd/back-ldbm/ldbm_search.c:
```diff
@@ -1695,31 +1694,41 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
+                     * we need to STILL apply the filter test with "as executed" in case of a test threshold
+                     * shortcut (lest we accidentally prevent the user seeing what they wanted ....)
+                     */
+                    filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
```
To be honest I am a bit scared here. Evaluation of the filter is expensive, here instead of one filter matching we may do two matching. So the gain on candidate list may be lost into a loss in more expensive filter matching.

Firstyear replied:
The whole theory of the filter test threshold is that filter eval is faster than evaling idls because we don't have an idl cache so they are always IO bound. That's why we have the ability to shortcut out and then evaluate the remaining conditions of the filter on the entry rather than doing full-indexed queries and bypassing the filter test. 


I'm actually reading the current master branch code and I have no idea how this it actually working correctly for current queries. 

Looking at this we have `onelevel_candidates()` which wraps the filter as intended by the user, with `(&(filter_intended)(parendid=<id>))`.  Lets say the filter_intended is `uid=a`. We then call into filter_candidates, which will look into the first elements of the filter_indended, and see it's equality, we'll hit the test_threshold and then return out with an idlist of just uid=a,. 

But then in onelevel, we discard the wrapper filter, and return the flag for forcing the filter test. So we then filter test only on filter_intended. 

I'm trying to think of ways to corrupt this, because I think it may be possible to ask for a query that would have subtree elements accidentally included in our partial candidate set since we never actually evaluated parentid=, and we don't apply that condition in our filter test.


So that's why we MUST absolutely test the filter twice! We have to check "once" for the access controls of "what did the user intend" and a second of "What is the correct *query* that gets the right results". Both onelevel and referrals matter here because the user may not have read on objectClass or parentid, and because we inject both, we need to check these are properly in the result sets else we are excluding/over-including in our results.

tbordaz replied:
> The whole theory of the filter test threshold is that filter eval is faster than evaling idls because we don't have an idl cache so they are always IO bound. That's why we have the ability to shortcut out and then evaluate the remaining conditions of the filter on the entry rather than doing full-indexed queries and bypassing the filter test.

filter threshold is to interrupt the process of building (from the filter) the candidate list when the #candidate is low (10 IIRC) and move immediately to filter evaluation. IIRC the main issue with filter threshold is when we have a 'OR' that force to fetch candidates for all filter components.

tbordaz reviewed (commented):
The overall patch looks nice, however I have some minor comments.

Regarding index_subsystems.c looks huge and I do not know the purpose of this functions, could you document in the PR its purpose and the entry points ?

There is a large set of tests to result into filter optimisation. I would like to see some user case where, for a given filter, we can verify gains (throughput or CPU/IO rate or response time or...)

tbordaz reviewed (approved):
I failed to reproduce https://github.com/389ds/389-ds-base/pull/5171#issuecomment-1058753981. All etimes are below 1ms so it is uncertain to compare the numbers (ranging 0.16ms to 0.58ms). I tested with idlscanlimit - 4K  to100K

Thanks @Firstyear for your great explanations. My concern was to get similar etimes when the idl threshold (10) was missed and that is the case.

Sorry for this long discussion I am happy to ack your patch.

Firstyear committed:
Issue 5170 - RFE - Filter optimiser

Bug Description: This introduces a query optimiser to 389-ds.
A query optimiser helps our backend process queries in a manner
that is more efficient by reducing IO events, and returning
the correct results quicker. This is needed as most applications
are not aware of the internals of how we execute a query internally
 - nor should they need to know. As an example, queries commonly
produced by SSSD are suboptimial and cause high latency and IO, but
this optimiser can resolve the majority of these issues which
generally improves all server throughput.

Fix Description: The optimiser works by attempting to produce
"the smallest candidate set" first. Having a smaller candidate
set earlier means that we can be below the filter test threshold
which shortcuts and can evaluate the remaining filter assertions
on the partial candidate set, rather than loading more expensive
and larger indexes. This also allows us to "fail fast" by
promoting "unlikely" candidates earlier which may end up
having a zero-length so that again, we don't load larger indexes.

A number of other changes were made and resolved in this change:

* Issue: The onelevel filter (not related to the
    optimiser ...) works be injecting a parentid field check
    to the filter that is used in the search. If the user
    does not have access to read this field, then one level
    searches fail

* Fix: Split the filter by "intent" and "as
    executed". We can filter test on as executed without check
    of the ACI, but we also check the filter as executed to
    ensure that only valid entries matching the proper semantics
    are returned

* Fix: Duplicate the filter for optimisation instead
    and free it in the search path, leaving the pblock filter as
    the "query as intended".

* Issue: In some cases, the filter test shortcut was
    applied, but with a specially crafted query, this could trigger
    the return prematurely giving incorrect results as the filter
    test was NOT bubbled up correctly.

* Fix: Flag in the sr that the filter test MUST be
    applied in certain shortcut cases.

fixes: https://github.com/389ds/389-ds-base/issues/5170

Author: William Brown <william@blackhats.net.au>

Review by: @tbordaz, @mreynolds389
